docs/design/services/s3/transfermanager/prototype.java

```diff
@@ -0,0 +1,622 @@
+                        .presignedUrl(presignedUrl)
+                        .build(),
+                file);
+    }
+
+    /**
+     * UploadObject a file to S3.
+     */
+    default UploadObject uploadObject(UploadObjectRequest request, Path file) {
+        return uploadObject(request, ObjectPartStreamCreator.forFile(file));
+    }
+
+    /**
+     * UploadObject an object to S3, using {@code partStreamFactory} to provide
+     * stream(s) for its part(s).
+     * <p>
+     * This is intended for advanced use not covered by other overloads of
+     * {@code uploadObject} as implementing a {@code Publisher<ByteBuffer>} can
+     * be difficult and error prone. When possible, prefer the other overloads
+     * of {@code downloadObject}, or implementations of {@link
+     * ObjectPartStreamCreator} provided by the SDK.
+     * <p>
+     * <b>Usage Example</b>
+     * <pre>
+     * UploadObject myUpload = tm.uploadObject(UploadObjectRequest
+     *         .builder()
+     *         .presignedUrl(url)
+     *         .build(),
+     *         ObjectPartStreamFactory.forInputStream(myStream));
+     * myUpload.completionFuture().join();
+     * </pre>
+     *
+     * @param request The upload request.
+     * @param partStreamFactory The part stream factory.
+     */
+    UploadObject uploadObject(UploadObjectRequest request, ObjectPartStreamCreator partStreamFactory);
+
+    /**
+     * Resume a previously paused object upload.
+     */
+    UploadObject resumeUploadObject(UploadObjectState uploadObjectState);
+
+    /**
+     * Create an {@code S3TransferManager} using the default values.
+     */
+    static S3TransferManager create() {
+        return builder().build();
+    }
+
+    static S3TransferManager.Builder builder() {
+        return ...;
+    }
+
+    interface Builder {
+        /**
+         * The custom S3AsyncClient this transfer manager will use to make calls
+         * to S3.
+         */
+        Builder s3client(S3AsyncClient s3Client);
+
+        /**
+         * The configuration for this transfer manager.
+         */
+        Builder configuration(TransferManagerConfiguration configuration);
+
+        /**
+         * Add a progress listener to the currently configured list of
+         * listeners.
+         */
+        Builder addProgressListener(TransferProgressListener progressListener);
+
+        /**
+         * Set the list of progress listeners, overwriting any currently
+         * configured list.
+         */
+        Builder progressListeners(Collection<? extends TransferProgressListener> progressListeners);
+
+        S3TransferManager build();
+    }
+}
+
+/**
+ * Configuration object for {@link S3TransferManager}.
+ */
+public interface TransferManagerConfiguration {
+    /**
+     * The aggregate max upload rate in bytes per second over all active upload
+     * transfers. The default value is {@link Long#MAX_VALUE}, or effectively
+     * unlimited.
+     */
+    long maximumUploadBytesPersSecond();
```

Comment:
I wonder if it's worth separating download and upload configurations. Currently, there are just too many configurations and we will potentially add more in the future.

Reply:
Yes +1 to this, I was thinking the same thing yesterday after going through this again.